src/fairseq2/recipes/lm/preference_finetune.py

```diff
@@ -0,0 +1,439 @@
+from fairseq2.data.text import load_text_tokenizer
+from fairseq2.datasets import LengthBatching
+from fairseq2.datasets.preference import (
+    GenericPreferenceOptimizationDataset,
+    PreferenceOptimizationBatch,
+    load_preference_optimization_dataset,
+)
+from fairseq2.logging import get_log_writer
+from fairseq2.models import load_model
+from fairseq2.models.decoder import DecoderModel
+from fairseq2.nn.checkpointing import use_layerwise_activation_checkpointing
+from fairseq2.nn.transformer import enable_memory_efficient_torch_sdpa
+from fairseq2.nn.utils.module import freeze_parameters
+from fairseq2.optim import AdamW
+from fairseq2.optim.lr_scheduler import CosineAnnealingLR
+from fairseq2.recipes.lm.preference_units.dpo_unit import (
+    DpoFinetuneConfig,
+    DpoFinetuneUnit,
+)
+from fairseq2.recipes.lm.preference_units.preference_criterion_config import (
+    PreferenceCriterionConfig,
+)
+from fairseq2.recipes.trainer import AbstractTrainUnit, Trainer
+from fairseq2.recipes.utils.asset import retrieve_asset_card
+from fairseq2.recipes.utils.log import log_model
+from fairseq2.recipes.utils.setup import compile_model, setup_gangs, to_data_parallel
+from fairseq2.typing import META, DataType
+from fairseq2.utils.profiler import Stopwatch
+
+log = get_log_writer(__name__)
+
+
+@dataclass
+class PreferenceOptimizationConfig:  # TODO: Should this just inherit from InstructionFinetuneConfig? The potential reason not to is that a later version may take two datasets (one positive, one negative)?
+    """Holds the configuration of a language model preference-finetuning task."""
+
+    # Data
+    dataset: Union[str, Path] = "openeft"  # TODO: change!
+    """The name, path, or path to the asset card of the preference optimization dataset."""
+
+    max_seq_len: int = 8192
+    """The maximum sequence length."""
+
+    max_num_tokens: int = 8192 * 2
+    """The maximum number of tokens per batch."""
+
+    example_shuffle_window: int = 10_000
+    """The size of the sliding window for shuffling examples."""
+
+    batch_shuffle_window: int = 1000
+    """The size of the sliding window for shuffling batches."""
+
+    num_prefetch: int = 4
+    """The number of batches to prefetch in background."""
+
+    # Criterion
+    criterion: str = "dpo"
+    """The type of preference optimization to perform"""
+
+    criterion_config: PreferenceCriterionConfig = field(
+        default_factory=lambda: DpoFinetuneConfig()
+    )  # TODO: is there a better way to do this?
+    """The hyperparameters specific to the criterion_type"""
+
+    # Model
+    model: Union[str, Path] = "llama3_8b_instruct"
+    """The name or path to the asset card of the language model to finetune."""
+
+    dtype: DataType = torch.bfloat16
+    """The data type of the model."""
+
+    data_parallelism: Literal[
+        "ddp", "fsdp"
+    ] = "fsdp"  # TODO is it fine to assume the reference model will use the same data_parallelism?
```

Comment:
In fact, reference model does not even need data parallelism. We should "fix" it in this recipe.